### PR TITLE
Save some bandwidth and allow Zxcvbn to be loaded asynchronously

### LINKED ISSUE #1: How can we async load zxcvbn?
Would be nice to have a way to load this script asynchronously with the app rather than bundling it at compile time, since it's a hefty 3-600 kb (depending on gzip vs uncompressed).
## 

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27741630-how-can-we-async-load-zxcvbn?utm_campaign=plugin&utm_content=tracker%2F25038149&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F25038149&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by jrjohnson**
By popular demand from #68 adds passwordStrength service and deprecates the shim.

The password strength service gives us the ability to use Zxcvbn
asynchronously and load it only when needed saving significant and often
unnecessary space in our vendor.js file.

Adds a build option bundleZxcvbn which defaults to true to maintain the
current behavior. When this option is set to false zxcvbn will not be
bundled into the apps vendor.js file.

There is one bit of strangeness to this. Despite hard coding the path to `zxcvbn.js` in my tests with a production build `zxcvbn.js` was fingerprinted and the path was updated automatically. I hope this is just ember being awesome, but since I don't understand exactly why it worked it would be nice to test this in other production like builds. cc/ @JoshSmith @st-h 

Fixes #1

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -7,22 +7,50 @@ var MergeTrees = require('broccoli-merge-trees');
...
-    app.import('vendor/shims/password-strength.js');
+
+    var addonOptions = (this.parent && this.parent.options) || (this.app && this.app.options) || {};
+    this.passwordStrengthConfig = addonOptions[this.name] || {
+      bundleZxcvbn: true
+    };
+
+    if (this.passwordStrengthConfig.bundleZxcvbn) {
+      app.import('vendor/zxcvbn.js');
+      app.import('vendor/shims/password-strength.js');
```
Comment (elwayman02) [inline index.js]: Nice job making this backwards compatible even though the default is a BC!

**Comment by elwayman02** [comment]
I like it!

**Comment by st-h** [comment]
Great. This is quite nice, but I had something a little different in mind. I changed the strength method to return an object which is able to deal with the promise and can be resolved within a template. This allows you to do something like that within a controller
```
strength: function() {
  return this.get('passwordStrength').strength(this.get('password'));
}.property('password'),
```
and in your template
```
<div>{{strength.crack_times_display.online_no_throttling_10_per_second}}</div>
```

This should be quite usefull if you only need to get simple things like the score. I like that quite more than having to deal with POJOs and handle the promise yourself.

Have a look here: https://github.com/st-h/ember-cli-password-strength/blob/6982006d5d3c8c4b1e8095f9efcf9c27fdfe689d/addon/services/password-strength.js#L39-L48

**Comment by jrjohnson** [comment]
@st-h thanks for the input.  In our app we prefer to deal with an unproxied promises since we use async/await syntax in our JS files and ember-promise-helpers in our templates. This allows for something like: 

```javascript
passwordStrengthScore: computed('password', async function () {
    const passwordStrength = this.get('passwordStrength');
    const obj = await passwordStrength.strength(this.get('password'));
    return obj.score;
  }),
```

and in the template

```handlebars
<meter max="4" value={{await passwordStrengthScore}}></meter>
```

I would prefer to leave wrapping the POJO response of zxcvbn into a PromiseProxy to individual application authors and not make it the response of this library so that we support both cases.

**Comment by st-h** [comment]
Ah, I see. Just out of curiousity.. What do you gain by doing so? And wouldn't you still be able to use async/await syntax when a proxied promise is returned?

**Comment by elwayman02** [comment]
@jrjohnson could the password-strength service somehow expose both options to consumers?

**Comment by jrjohnson** [comment]
Yes. Ok. I have added a third method to the service `strengthProxy` returns a promise wrapped in a proxy object. I also added some docs to the service. I'm not sure IDEs can read these, but it does make the code a bit easier to grok.

**Comment by jrjohnson** [comment]
@st-h does the `strengthProxy` method meet you needs? If so I think this is ready to merge.

**Comment by st-h** [comment]
@jrjohnson lgtm 👍  Did a few quick tests as well and everything seems to work as it should. Nice idea adding both.

On another note: How do you guys feel about adding a template helper `{{password-score password=...}}` which already offers a fully styled indicator to add to your app? However, this could easily be a dedicated addon which makes use of this addon if you prefer to keep everything as light as possible.

**Comment by elwayman02** [comment]
I'd be in support of keeping this lightweight, as every app will have its own needs for the score and may not want to inherit any styles. Making a separate addon on top of this sounds great, though!

**Comment by st-h** [comment]
@elwayman02 never mind. looks like somebody already took care of that https://www.npmjs.com/package/ember-paper-password

**Comment by elwayman02** [comment]
That's awesome, @st-h! Though they will probably need a PR to handle the breaking changes in this PR once it's released.

**Pull request merged**